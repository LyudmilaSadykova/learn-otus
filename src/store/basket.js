import { defineStore } from "pinia";
import { ref } from "vue";

export const useBasketStore = defineStore('basket', () => {
    const basketList = ref([])

    function addToBasket(product) {
        basketList.value.push(product);
    }

    function clearBasket() {
        basketList.value = [];
    }

    function delToBasket(id) {
        basketList.value = JSON.parse(JSON.stringify(basketList.value.filter((item) => item.id !== id)));
    }

    return { basketList, addToBasket, clearBasket, delToBasket }
})