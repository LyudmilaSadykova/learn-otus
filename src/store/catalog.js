import { defineStore } from "pinia";
import { ref, computed } from "vue";
import axios from "axios";

export const useCatalogStore = defineStore('catalog', () => {
    const loading = ref(false);
    const productList = ref([])
    const filters = ref([])

    async function loadProducts() {
        loading.value = true;

        await axios.get('https://fakestoreapi.com/products')
        .then(response => {
            productList.value = response.data
            //console.log(productList.value);
        })
        .catch(error => {
            console.log(error);
        })
        .finally(() => loading.value = false)
    }

    const productListHome = computed(() => [...productList.value].splice(0, 4))

    function updateFilters(filterList) {
        filters.value = filterList;
    }

    const productListFiltered = computed(() => {
        return [...productList.value].filter((product) => 
            ((!filters.value[0] || filters.value[0] && product.title.indexOf(filters.value[0]) > -1) &&
            (!filters.value[1] || filters.value[1] && product.price > filters.value[1]) && (!filters.value[2] || filters.value[2] && product.price < filters.value[2]))
          );
    })

    function addProduct(product) {
        productList.value.push(product);
    }

    return { loading, productList, loadProducts, productListHome, updateFilters, productListFiltered, addProduct }
})