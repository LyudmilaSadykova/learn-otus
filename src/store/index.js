import { createStore } from 'vuex';
import axios from 'axios';

export default createStore({
    state: {
        productList: [],
        filters: [],
        basketList: []
    },
    getters: {
        getProducts(state) {
          return state.productList;
        },
        filteredProducts(state) {
            return [...state.productList].filter((product) => 
                ((!state.filters[0] || state.filters[0] && product.title.indexOf(state.filters[0]) > -1) &&
                (!state.filters[1] || state.filters[1] && product.price > state.filters[1]) && (!state.filters[2] || state.filters[2] && product.price < state.filters[2]))
              );
        },
        getBasketList(state) {
            return state.basketList;
        }
    },
    mutations: {
        SET_PRODUCTS(state, products) {
            state.productList = products;
        },
        PUSH_PRODUCT(state, product) {
            state.productList.push(product);
        },
        SET_FILTERS(state, filterList) {
            state.filters = filterList;
        },
        PUSH_BASKET(state, product) {
            state.basketList.push(product);
        },
        SET_BASKET(state) {
            state.basketList = [];
        },
    },
    actions: {
        /*async loadProducts({ commit }) {
            await axios.get('https://fakestoreapi.com/products')
            .then(response => {
                commit('SET_PRODUCTS', response.data)
                //console.log(productList.value);
            })
            .catch(error => {
                console.log(error);
            })
        },*/
        updateProducts({ commit }, productList) {
            commit('SET_PRODUCTS', productList)
        },
        addProduct({ commit }, product) {
            commit('PUSH_PRODUCT', product)
        },
        updateFilters({ commit }, filterList) {
            commit('SET_FILTERS', filterList)
        },
        addToBasket({ commit }, product) {
            commit('PUSH_BASKET', product)
        },
        clearBasket({ commit }) {
            commit('SET_BASKET')
        },
    }
    
})