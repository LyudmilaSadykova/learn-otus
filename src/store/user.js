import { defineStore } from "pinia";
import { ref } from "vue";

export const useAccountStore = defineStore('account', () => {
    const accountData = ref([])
    const isAuth = ref(false)

    function saveData(data) {
        accountData.value = data;
    }

    function loginUser(token) {
        localStorage.setItem('token', token);
        isAuth.value = true;
    }

    function logoutUser() {
        localStorage.removeItem('token');
        isAuth.value = false;
    }

    return { accountData, isAuth, saveData, loginUser, logoutUser }
})