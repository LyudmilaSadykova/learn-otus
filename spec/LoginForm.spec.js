import { mount } from '@vue/test-utils';
import component from '../src/view/LoginForm.vue';
import { createTestingPinia } from '@pinia/testing';
import { useAccountStore } from '../src/store/user';
import { vi } from 'vitest';


describe('LoginForm component', () => {
    const buildWrapper = () => mount(component, {
        global: {
            plugins: [createTestingPinia({
                createSpy: vi.fn,
                stubActions: false,
                initialState: {
                    isAuth: false
                }}
            )],
        },
    });

    it('mount without errors', () => {
        const wrapper = buildWrapper();
        expect(wrapper.exists()).toBe(true);
        expect(wrapper.element).toMatchSnapshot();
    });

    it('find login header', () => {
        const wrapper = buildWrapper();
        expect(wrapper.text()).contain('Войти в аккаунт');
    });

    it('find login button',() => {
        const wrapper = buildWrapper();
        const button = wrapper.find('button');
        expect(button.exists()).toBe(true);
    });

    it('emit input fields', async () => {
        const wrapper = buildWrapper();

        const inputLogin = wrapper.find('[name="login"]');
        expect(inputLogin.exists()).toBe(true);
        await inputLogin.setValue('admin');

        const inputPassword = wrapper.find('[name="password"]');
        expect(inputPassword.exists()).toBe(true);
        await inputPassword.setValue('123456')

        // const button = wrapper.find('button');
        // await button.trigger('click');

        const account = useAccountStore()
        expect(account.isAuth).toBe(false)
        
        account.loginUser('admin123456')
        expect(account.loginUser).toHaveBeenCalledTimes(1)
        expect(account.isAuth).toBe(true)

        expect(wrapper.text()).contain('Заполните данные профиля для оформления заказа');
    });
});