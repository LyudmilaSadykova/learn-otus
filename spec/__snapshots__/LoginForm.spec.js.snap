// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginForm component > mount without errors 1`] = `
<div
  data-v-app=""
>
  
  <h1
    data-v-366a8352=""
  >
    Войти в аккаунт
  </h1>
  <!--v-if-->
  <form
    data-v-366a8352=""
    novalidate=""
  >
    <v-row
      data-v-366a8352=""
    >
      <v-col
        class="left-col"
        cols="5"
        data-v-366a8352=""
      >
        Логин
      </v-col>
      <v-col
        cols="7"
        data-v-366a8352=""
      >
        <input
          data-v-366a8352=""
          name="login"
          type="string"
        />
        <!---->
      </v-col>
    </v-row>
    <v-row
      data-v-366a8352=""
    >
      <v-col
        class="left-col"
        cols="5"
        data-v-366a8352=""
      >
        Пароль
      </v-col>
      <v-col
        cols="7"
        data-v-366a8352=""
      >
        <input
          data-v-366a8352=""
          name="password"
          type="password"
        />
        <!---->
      </v-col>
    </v-row>
    <v-row
      data-v-366a8352=""
    >
      <v-col
        class="search-btn"
        cols="12"
        data-v-366a8352=""
      >
        <button
          data-v-366a8352=""
        >
           Войти 
        </button>
      </v-col>
    </v-row>
  </form>
  <div
    class="valid_message"
    data-v-366a8352=""
    style="display: none;"
  >
    Заполните данные профиля для оформления заказа
  </div>
  
</div>
`;
